analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDBExtension.java

```diff
@@ -0,0 +1,325 @@
+        } else if (forge.equals("PyPI")) {
+            return new ExtendedRevisionPythonCallGraph(json);
+        }
+
+        return null;
+    }
+
+    /**
+     * Saves a callgraph of new format to the database to appropriate tables.
+     *
+     * @param callGraph   Call graph to save to the database.
+     * @param metadataDao Data Access Object to insert records in the database
+     * @return Package ID saved in the database
+     */
+    protected long saveToDatabase(ExtendedRevisionCallGraph callGraph, MetadataDao metadataDao) {
+        // Insert package record
+        final long packageId = metadataDao.insertPackage(callGraph.product, callGraph.forge);
+
+        // Insert package version record
+        final long packageVersionId = metadataDao.insertPackageVersion(packageId,
+                callGraph.getCgGenerator(), callGraph.version, null,
+                getProperTimestamp(callGraph.timestamp), new JSONObject());
+
+        var callables = insertDataExtractCallables(callGraph, metadataDao, packageVersionId);
+        var numInternal = callables.size();
+
+        var callablesIds = new LongArrayList(callables.size());
+        // Save all callables in the database
+        callablesIds.addAll(metadataDao.insertCallablesSeparately(callables, numInternal));
+
+        // Build a map from callable Local ID to Global ID
+        var lidToGidMap = new Long2LongOpenHashMap();
+        for (int i = 0; i < callables.size(); i++) {
+            lidToGidMap.put(callables.get(i).getId().longValue(), callablesIds.getLong(i));
+        }
+
+        // Insert all the edges
+        var edges = insertEdges(callGraph.getGraph(), lidToGidMap, metadataDao);
+
+        // Remove duplicate nodes
+        var internalIds = new LongArrayList(numInternal);
+        var externalIds = new LongArrayList(callablesIds.size() - numInternal);
+        for (int i = 0; i < numInternal; i++) {
+            internalIds.add(callablesIds.getLong(i));
+        }
+        for (int i = numInternal; i < callablesIds.size(); i++) {
+            externalIds.add(callablesIds.getLong(i));
+        }
+        var internalNodesSet = new LongLinkedOpenHashSet(internalIds);
+        var externalNodesSet = new LongLinkedOpenHashSet(externalIds);
+        numInternal = internalNodesSet.size();
+        callablesIds = new LongArrayList(internalNodesSet.size() + externalNodesSet.size());
+        callablesIds.addAll(internalNodesSet);
+        callablesIds.addAll(externalNodesSet);
+
+        // Create a GID Graph for production
+        this.gidGraph = new GidGraph(packageVersionId, callGraph.product, callGraph.version,
+                callablesIds, numInternal, edges);
+        return packageVersionId;
+    }
+
+    public ArrayList<CallablesRecord> insertDataExtractCallables(
+            ExtendedRevisionCallGraph callgraph, MetadataDao metadataDao, long packageVersionId) {
+        return new ArrayList<CallablesRecord>();
```

Comment:
I assume this is a TODO? :)

Reply:
This was added on #91 . @StefanosChaliasos could you please take a look?

Reply:
Nope, all classes that implement this class must provide an implementation for this method. Unfortunately, we cannot convert this class to an abstract class. I added a comment.